### PR TITLE
Br 1

### PR DESCRIPTION
**Pull request by MaliukDaria**

### CODE REVIEW

File: src/main/java/com/example/demo/controller/ReadFileController.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.example.demo.controller;
+
+import com.example.demo.service.file.CustomFileReader;
+import java.util.List;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
```
Comment (AnaLanda) [inline src/main/java/com/example/demo/controller/ReadFileController.java]: Are you sure this should be a controller, and not internal business logic?

File: src/test/java/com/example/demo/service/file/CustomFileReaderTest.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import java.io.PrintWriter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+import org.junit.Assert;
+import org.junit.jupiter.api.Test;
+
+public class CustomFileReaderTest {
+
+    @Test
```
Comment (AnaLanda) [inline src/test/java/com/example/demo/service/file/CustomFileReaderTest.java]: I think you could add a couple tests for bad scenarios, e.g. trying to read a non-existent or empty file)

**Review by AnaLanda — COMMENTED** [review]
Looks good overall, but I'm not sure that the file reading logic should be implemented as a controller)

File: src/test/java/com/example/demo/service/file/CustomFileReaderTest.java (hunk 2 above)
Comment (MaliukDaria) [reply src/test/java/com/example/demo/service/file/CustomFileReaderTest.java]: agree

File: .gitignore [hunk 3]
```diff
@@ -34,3 +34,4 @@ build/
 
 .idea/*
 target/*
+/src/main/resources/reviews.csv
```
Comment (KseniiaMakarova) [inline .gitignore:35]: I can see .idea/ and target/ already indicated above
let's maybe add `logs` folder?

File: src/main/java/com/example/demo/service/file/CustomFileReader.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.log4j.Logger;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CustomFileReader {
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/demo/service/file/CustomFileReader.java]: this class implements reading file from file system

we can create some abstract FileReader interface and make this class implement it (and call it FileSystemFileReader, for example)

and then use the abstract type for injection

File: src/test/java/com/example/demo/service/file/CustomFileReaderTest.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        } finally {
+            file.delete();
+        }
+        Assert.assertEquals(actual.size(), 0);
+    }
+
+    @Test
+    public void readNonExistentFile() {
+        Assertions.assertThrows(
+                RuntimeException.class, () -> fileReader.readFile("test.csv"));
```
Comment (KseniiaMakarova) [inline src/test/java/com/example/demo/service/file/CustomFileReaderTest.java]:
```suggestion
                RuntimeException.class, () -> fileReader.readFile(FILE_PATH));
```

File: src/main/java/com/example/demo/controller/ReadFileController.java (hunk 1 above)
Comment (KseniiaMakarova) [reply src/main/java/com/example/demo/controller/ReadFileController.java]: after the discussion with Bohdan, I would say that if you want to 100% control the time when you want to load data - then launching this process from controller is ok

but probably you will like the ApplicationListener more
https://reflectoring.io/spring-boot-execute-on-startup/#applicationlistener

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/demo/service/file/CustomFileReader.java (hunk 4 above)
Comment (MaliukDaria) [reply src/main/java/com/example/demo/service/file/CustomFileReader.java]: done

File: src/test/java/com/example/demo/service/file/CustomFileReaderTest.java (hunk 5 above)
Comment (MaliukDaria) [reply src/test/java/com/example/demo/service/file/CustomFileReaderTest.java]: done

File: .gitignore (hunk 3 above)
Comment (MaliukDaria) [reply .gitignore:35]: done

File: src/main/java/com/example/demo/service/ReviewsApplicationListener.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    public ReviewsApplicationListener(CustomFileReader fileReader) {
+        this.fileReader = fileReader;
+    }
+
+    @Override
+    public void onApplicationEvent(ApplicationReadyEvent applicationReadyEvent) {
+        LOGGER.info("MyApplicationListener: onApplicationEvent method starts");
+        Properties fileProps = new Properties();
+        try {
+            fileProps.load(new FileInputStream("src/main/resources/sourсeFile.properties"));
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/demo/service/ReviewsApplicationListener.java]: let's just create a field in this class instead which value would be taken automatically from application.properties
see https://www.baeldung.com/spring-value-annotation

File: src/main/java/com/example/demo/service/ReviewsApplicationListener.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        } catch (IOException e) {
+            LOGGER.error("Cant load sourсeFile.properties file", e);
+            throw new RuntimeException("Cant load sourсeFile.properties file", e);
+        }
+        List<String> fileData =
+                fileReader.readFile(fileProps.getProperty("sourceFilePath"));
+        System.out.println();
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/demo/service/ReviewsApplicationListener.java]: let's remove this line

File: src/main/resources/sourсeFile.properties [hunk 8]
```diff
@@ -0,0 +1 @@
+sourceFilePath=src/main/resources/reviews.csv
```
Comment (KseniiaMakarova) [inline src/main/resources/sourсeFile.properties]: move this to application.properties

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/demo/service/ReviewsApplicationListener.java (hunk 6 above)
Comment (MaliukDaria) [reply src/main/java/com/example/demo/service/ReviewsApplicationListener.java]: done

File: src/main/java/com/example/demo/service/ReviewsApplicationListener.java (hunk 7 above)
Comment (MaliukDaria) [reply src/main/java/com/example/demo/service/ReviewsApplicationListener.java]: done

File: src/main/resources/sourсeFile.properties (hunk 8 above)
Comment (MaliukDaria) [reply src/main/resources/sourсeFile.properties]: done

**Review by KseniiaMakarova — APPROVED** [review]
good job!

File: src/main/java/com/example/demo/service/ReviewsApplicationListener.java [hunk 9]
```diff
@@ -1,19 +1,21 @@
...
-import java.util.Properties;
 import org.apache.log4j.Logger;
+import org.springframework.beans.factory.annotation.Value;
 import org.springframework.boot.context.event.ApplicationReadyEvent;
 import org.springframework.context.ApplicationListener;
+import org.springframework.context.annotation.PropertySource;
 import org.springframework.stereotype.Component;
 
 @Component
+@PropertySource("classpath:application.properties")
```
Comment (KseniiaMakarova) [inline src/main/java/com/example/demo/service/ReviewsApplicationListener.java]: this may work without this line as well

**Pull request merged**